### ISSUE TITLE
TFP backend - LKJ parameter shapes

### ISSUE THREAD

**Issue by adamhaber** [stan-dev/stanc3#407]
The following program:
```stan
data {
  int K;
}
parameters {
  cholesky_factor_corr[K] L;
}
model {
   L ~ lkj_corr_cholesky(2.0);
}
```
Produces the following parameter_shapes method:
```python
def parameter_shapes(self, nchains__):
    K = self.K
    return [(nchains__, K * K - tf__.cast(1, tf__.float64) / tf__.cast(2, tf__.float64))]
```
I assume this should be K * (K-1) / 2?

**Comment by nhuurre** [comment]
Looks like the code gen adds casts too often. We could change this
https://github.com/stan-dev/stanc3/blob/c63095d306f45b97364d985950b8cfd62509a7c9/src/tfp_backend/Code_gen.ml#L14-L17
to cast only `Lit (Real, s)` and not `Lit (Int, s)`. I don't know what side effects that has. The problem is that one can write an integer literal where the code expects a real number. I assume we need a cast in that case. Maybe we need to first implement stan-dev/stanc3#373.

**Comment by adamhaber** [comment]
You're right - it seems like stan-dev/stanc3#411 fixed the parenthesis issue, and now there's just the extra casts.

**Comment by adamhaber** [comment]
I guess we can close this one as well? I don't think the `tf__.cast` is actually problematic, but I might be missing something. Perhaps we're OK with the auto-casting but we prefer to cast to int instead?

**Comment by brianwa84** [comment]
Presumably parameter_shapes is used to initialize a chain as the shape argument to something like `tf.random.uniform(shape=param_shp, minval=-2, maxval=2.)` In this case float tensors won't work. You could consider using `_default_event_space_bijector()` on `CholeskyLKJ` (and others) to pull back to an unconstrained shape.

Something like:
```python
nchains = 7
@tf.function
def unconstrained_shape():
  d = tfp.distributions.CholeskyLKJ(5, tf.cast(.5, tf.float64))
  return d._default_event_space_bijector().inverse(d.sample(nchains))
f.get_concrete_function().structured_outputs
```
(Using `tf.function`/`get_concrete_function` avoids actually materializing a possibly big sample, and just traces the execution.)

**Comment by adamhaber** [comment]
> Presumably parameter_shapes is used to initialize a chain as the shape argument to something like tf.random.uniform(shape=param_shp, minval=-2, maxval=2.) In this case float tensors won't work. 

@brianwa84 casting the shape arguments to tf.int32 seems to work (see stan-dev/stanc3#592). What do you think?
 
>You could consider using _default_event_space_bijector() on CholeskyLKJ (and others) to pull back to an unconstrained shape. 

The shapes themselves are already handled AFAIK, it's only their dtypes that were problematic.

> to cast only Lit (Real, s) and not Lit (Int, s).

I've tried this, but this creates a dtypes mess - for example, if chaining Shift(a) (which has default dtype of float32) and Scale(tf.cast(b, tf.float64)) (which would happen if b is real and therefore would be casted as tf.float64) fails due to dtype mismatch.

**Comment by brianwa84** [comment]
int32 should work, though it would also be valid to leave these as raw python int.
